src/SchemaReaderCallbackAbstraction.php

```diff
@@ -0,0 +1,216 @@
+<?php
+namespace GoetasWebservices\XML\XSDReader;
+
+use Closure;
+use DOMDocument;
+use DOMElement;
+use DOMNode;
+use DOMNodeList;
+use GoetasWebservices\XML\XSDReader\Exception\IOException;
+use GoetasWebservices\XML\XSDReader\Exception\TypeException;
+use GoetasWebservices\XML\XSDReader\Schema\Attribute\Attribute;
+use GoetasWebservices\XML\XSDReader\Schema\Attribute\AttributeDef;
+use GoetasWebservices\XML\XSDReader\Schema\Attribute\AttributeItem;
+use GoetasWebservices\XML\XSDReader\Schema\Attribute\Group as AttributeGroup;
+use GoetasWebservices\XML\XSDReader\Schema\Element\Element;
+use GoetasWebservices\XML\XSDReader\Schema\Element\ElementContainer;
+use GoetasWebservices\XML\XSDReader\Schema\Element\ElementDef;
+use GoetasWebservices\XML\XSDReader\Schema\Element\ElementItem;
+use GoetasWebservices\XML\XSDReader\Schema\Element\ElementRef;
+use GoetasWebservices\XML\XSDReader\Schema\Element\Group;
+use GoetasWebservices\XML\XSDReader\Schema\Element\GroupRef;
+use GoetasWebservices\XML\XSDReader\Schema\Element\InterfaceSetMinMax;
+use GoetasWebservices\XML\XSDReader\Schema\Exception\TypeNotFoundException;
+use GoetasWebservices\XML\XSDReader\Schema\Inheritance\Base;
+use GoetasWebservices\XML\XSDReader\Schema\Inheritance\Extension;
+use GoetasWebservices\XML\XSDReader\Schema\Inheritance\Restriction;
+use GoetasWebservices\XML\XSDReader\Schema\Item;
+use GoetasWebservices\XML\XSDReader\Schema\Schema;
+use GoetasWebservices\XML\XSDReader\Schema\SchemaItem;
+use GoetasWebservices\XML\XSDReader\Schema\Type\BaseComplexType;
+use GoetasWebservices\XML\XSDReader\Schema\Type\ComplexType;
+use GoetasWebservices\XML\XSDReader\Schema\Type\ComplexTypeSimpleContent;
+use GoetasWebservices\XML\XSDReader\Schema\Type\SimpleType;
+use GoetasWebservices\XML\XSDReader\Schema\Type\Type;
+use GoetasWebservices\XML\XSDReader\Utils\UrlUtils;
+use RuntimeException;
+
+abstract class SchemaReaderCallbackAbstraction extends AbstractSchemaReader
```

Comment:
`AbstractSchemaReader` is used only once... and is also abstract...why having it?

Reply:
same for `SchemaReaderCallbackAbstraction`

Reply:
same for all the other classes in the hierarchy

Reply:
SchemaReader's "F" rating was mostly due to the combined length of the methods. Splitting the class up was an slightly artificial way to bring the quality up, whilst also highlighting potential for proper abstraction (like there were a few methods that dealt solely with a particular class, so I refactored them onto that particular class).

Reply:
Inheritance IMO should be used when there is an "Inheritance relation between classes"... in this case I would consider using "composition". 

Im a promoter of [Composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)